README.md
LP55231 Linux Rust Driver
-------------------------

[![Open in Dev Containers](https://img.shields.io/static/v1?label=Dev%20Containers&message=Open&color=blue&logo=visualstudiocode)](https://vscode.dev/redirect?url=vscode://ms-vscode-remote.remote-containers/cloneInVolume?url=https://github.com/neurosity/ti-lp55231)

Linux driver for [Texas Instruments LP55231](https://www.ti.com/product/LP55231),
a 9 channel RGB/White LED controller with internal program memory and integrated
charge Pump.

**Features:**
- Full implementation of I2C control interface in [datasheet](https://www.ti.com/lit/ds/symlink/lp55231.pdf)
- Ergonomic API to leverage the [programming engine](#example-effect-blinking)
- Easy to debug with optional features:
    - [read-after-write checks](#read-after-write-verifications) to validate register writes
    - [debug output](#debug-output) to see the value in a register before and after a write

## Initialization, setup, and preparing animations

This example covers a typical initialization of the driver, preparing three
LEDs (R, G, B) for animated effects.

```rust
use ti_lp55231::{
  Channel,
  ChargePumpMode,
  ClockSelection,
  Direction,
  Engine,
  EngineExec,
  EngineMode,
  Instruction,
  LP55231,
  PreScale,
}

// Create the driver
let path = "/dev/i2c-2";
let i2c_addr = 0x32;
let ic = LP55231::create(path, i2c_addr)?;

// Power and configure the driver.
ic.set_enabled(true)?;
ic.set_misc_settings(Misc {
  auto_increment_enabled: true,
  powersave_enabled: true,
  charge_pump_mode: ChargePumpMode::Auto,
  pwm_powersave_enabled: true,
  clock_selection: ClockSelection::ForceInternal,
})?;

// Channel assignment.
let (r, g, b) = (Channel::D7, Channel::D1, Channel::D2);

// Enable logarithmic brightness for a smoother ramp up effect.
ic.set_log_brightness(r, true)?;
ic.set_log_brightness(g, true)?;
ic.set_log_brightness(b, true)?;

// Enable ratiometric dimming to preserve the ratio between the
// RGB components of all mapped channels during animations
ic.set_ratiometric_dimming(r, true)?;
ic.set_ratiometric_dimming(g, true)?;
ic.set_ratiometric_dimming(b, true)?;

// Set color to orange
ic.set_channel_pwm(r, 255)?;
ic.set_channel_pwm(g, 128)?;
ic.set_channel_pwm(b, 0)?;

// Program the IC (see other example for implementations of `create_program`)
let instructions = create_program(&[r, g, b])?;
ic.load_program(&instructions)?;

// Wait for the ENGINE_BUSY bit to clear,
// indicating that all instructions have been loaded.
ic.wait_while_engine_busy(Duration::from_millis(10))?;

// Set up one of the programming engines to Halt & Hold (ready to execute).
let engine = Engine::E1;
ic.set_engine_exec(engine, EngineExec::Hold)?;
ic.set_engine_mode(engine, EngineMode::Halt)?;

// Run the effect
ic.set_engine_exec(engine, EngineExec::Free)?;
ic.set_engine_mode(engine, EngineMode::RunProgram)?;
```

## Example effect: blinking

This example is an implementation of `create_program` that prepares a blinking
effect to run in an endless loop.

```rust
fn create_program(channels_to_control: &[Channel]) -> [Instruction; 8] {
  [
    // ----- LED-to-Engine mapping table
    // 00. Map all target output channels to the programming engine for control.
    Instruction::map_channels(channels_to_control),

    // ----- blink effect start
    // 01-02. Set LED mapping table start/end index + activation.
    Instruction::mux_map_start(0),
    Instruction::mux_ld_end(0),
    // 03. Power all mapped LEDs off.
    Instruction::set_pwm(0),
    // 04. Wait ~0.5 seconds (15.625ms * 30).
    Instruction::wait(PreScale::CT15_625, 30),
    // 05. Set all LEDs to max brightness.
    Instruction::set_pwm(255),
    // 06. Wait ~0.5 seconds (15.625ms * 30).
    Instruction::wait(PreScale::CT15_625, 30),
    // 07. Loop back to beginning of blink effect index.
    Instruction::branch(1, 0),
  ]
}
```

## Example effect: glow

```rust
fn create_program(channels_to_control: &[Channel]) -> [Instruction; 9] {
  [
    // ----- LED-to-Engine mapping table
    // 00. Map all target output channels to the programming engine for control.
    Instruction::map_channels(channels_to_control),

    // ----- glow effect start
    // 01-02. Set LED mapping table start/end index + activation.
    Instruction::mux_map_start(0),
    Instruction::mux_ld_end(0),
    // 03. Quickly ramp up to max brightness.
    Instruction::ramp(PreScale::CT0_488, 4, Direction::Up, 255),
    // 04. Wait ~0.5 seconds (15.625ms * 30 = 468.75ms).
    Instruction::wait(PreScale::CT15_625, 30),
    // 05. Begin ramping brightness down to half (255 - 127 = 128).
    Instruction::ramp(PreScale::CT15_625, 4, Direction::Down, 127),
    // 06. Wait ~0.5 seconds (15.625ms * 30 = 468.75ms).
    Instruction::wait(PreScale::CT15_625, 30),
    // 07. Begin ramping brightness up to max (128 + 127 = 255).
    Instruction::ramp(PreScale::CT15_625, 4, Direction::Up, 127),
    // 08. Loop back to first step of effect.
    Instruction::branch(1, 0),
  ]
}
```

## Switching between effects

The programming engine supports up to 96 instructions, which gives you plenty of
room to set up multiple effects. To switch between effects:

- pause the programming engine
- update the program counter to first index of next effect
- unpause the programming engine.

Example:

```rust
// Pause engine execution.
ic.set_engine_exec(Engine::E1, EngineExec::Hold)?;
ic.wait_while_engine_busy(Duration::from_millis(1))?;

// Update the program counter to the starting instruction of the desired effect
// This example assumes we're jumping to instruction 42, of the possible 96
// programming memory addresses.
ic.set_engine_program_counter(Engine::E1, 42)?;

// Unpause the engine and begin the new animation.
ic.set_engine_exec(Engine::E1, EngineExec::Free)?;
ic.set_engine_mode(Engine::E1, EngineMode::RunProgram)?;
```

# Debugging

## Read-after-write verifications

Read-after-write checks can be enabled with:

```rust
let ic = LP55231::create(...)?;
ic.verify_writes = true;
```

This will cause the driver to perform a read after every I2C write instruction
to compare the value in the register. It will throw an exception if the read
value does not match the written value.

> [!NOTE]
> This is useful during development, especially around using the programming
> engines which must be in the correct internal state in order to allow changes.

## Debug output

When enabled via `debug_enabled` property, the driver will emit useful (but
rather verbose) output to help you understand the state of registers with every
read and write operation. Example:

```rust
let ic = LP55231::create(...)?;
ic.debug_enabled = true;
ic.set_enabled(true)?;
```

Will produce output:

```
set_enabled(true) {
  00000000 << 0x00 ENABLE_ENGINE_CNTRL1
  00100000 >> 0x00 ENABLE_ENGINE_CNTRL1
}
```

### Scoping debug output for multiple I2C calls

Scope for multiple debug calls can be combined with the `debug::scope!` macro:

```rust
fn multiple_i2c_calls(
  ic: &mut LP55231,
  value: bool,
) -> Result<(), LinuxI2CError> {
  debug::scope!(ic, "example({})", value);
  ic.set_enabled(value)?;
  ic.set_enabled(!value)?;
  Ok(())
}

multiple_i2_calls(true)?;
```

Would result in the following output:
```
example(true) {
  set_enabled(true) {
    00000000 << 0x00 ENABLE_ENGINE_CNTRL1
    00100000 >> 0x00 ENABLE_ENGINE_CNTRL1
  }
  set_enabled(false) {
    00100000 << 0x00 ENABLE_ENGINE_CNTRL1
    00000000 >> 0x00 ENABLE_ENGINE_CNTRL1
  }
}
```

See [debug.rs](src/debug.rs) docs for more details.

## Getting started with development

1. Clone the project and open the folder in VS Code
2. Accept plugin suggestions (dev container required in non-linux envs)
3. Re-open in dev container

> [!NOTE]
> This project uses [hermit](https://cashapp.github.io/hermit/) to manage the
> Rust toolchain for this project. No prior installation of Rust required.

## TODO

- [ ] Read/write pages in blocks (`at_once` param in `read/write_program_page`)


CHANGELOG.md
# V1.0.0

First release!


src/mask.rs
use bitflags::bitflags;

use crate::{types::Engine, Channel};

bitflags! {
  #[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
  /// Masks.
  ///
  /// Sorted by register they apply to.
  /// *Not* a comprehensive list.
  ///
  /// Refer to spec sections 7.6.1 and 7.6.2
  pub struct Mask: u8 {
    // 00, ENABLE / ENGINE CNTRL1
    const CHIP_EN = 0b0100_0000;
    const ENGINE1_EXEC = 0b0011_0000;
    const ENGINE2_EXEC = 0b0000_1100;
    const ENGINE3_EXEC = 0b0000_0011;

    // 01 ENGINE CNTRL2
    const ENGINE1_MODE = 0b0011_0000;
    const ENGINE2_MODE = 0b0000_1100;
    const ENGINE3_MODE = 0b0000_0011;

    // 02 OUTPUT DIRECT/RATIOMETRIC MSB
    const D9_RATIO_EN = 0b0000_0001;

    // 03 OUTPUT DIRECT/RATIOMETRIC LSB
    const D8_RATIO_EN = 0b1000_0000;
    const D7_RATIO_EN = 0b0100_0000;
    const D6_RATIO_EN = 0b0010_0000;
    const D5_RATIO_EN = 0b0001_0000;
    const D4_RATIO_EN = 0b0000_1000;
    const D3_RATIO_EN = 0b0000_0100;
    const D2_RATIO_EN = 0b0000_0010;
    const D1_RATIO_EN = 0b0000_0001;

    // 04 OUTPUT ON/OFF CONTROL MSB
    const D9_ON = 0b0000_0001;

    // 05 OUTPUT ON/OFF CONTROL LSB
    const D8_ON = 0b1000_0000;
    const D7_ON = 0b0100_0000;
    const D6_ON = 0b0010_0000;
    const D5_ON = 0b0001_0000;
    const D4_ON = 0b0000_1000;
    const D3_ON = 0b0000_0100;
    const D2_ON = 0b0000_0010;
    const D1_ON = 0b0000_0001;

    // 06-0E, D1-D9 CONTROL
    const LOG_EN = 0b0010_0000;
    const MAPPING = 0b1100_0000;

    // 36 MISC
    const EN_AUTO_INCR = 0b0100_0000;
    const POWERSAVE_EN = 0b0010_0000;
    const CP_MODE = 0b0001_1000;
    const PWM_PS_EN = 0b0000_0100;
    const CLK_DET_EN = 0b0000_0011;

    // 3A, STATUS/INTERRUPT
    const LEDTEST_MEAS_DONE = 0b1000_0000;
    const MASK_BUSY = 0b0100_0000;
    const STARTUP_BUSY = 0b0010_0000;
    const ENGINE_BUSY = 0b0001_0000;
    const EXT_CLK_USED = 0b0000_1000;
    const ENG1_INT = 0b0000_0100;
    const ENG2_INT = 0b0000_0010;
    const ENG3_INT = 0b0000_0001;

    // 3D, RESET
    const RESET = 0b1111_1111;

    // 4F, PROG MEM PAGE SELECT
    const PAGE_SEL = 0b0000_0111;
  }
}

impl Mask {
  /// Apply the specified `value` using the [Mask] to `byte`.
  ///
  /// Example; given:
  /// - A mask `0b0000_1100`
  /// - A value `0b10`
  /// - A byte `0b1111_1111`
  /// Then:
  /// - `mask.apply(value, byte)` will produce `0b1111_1011`
  pub fn apply(&self, value: u8, to_byte: u8) -> u8 {
    let byte_with_mask_bits_cleared = to_byte & !self.bits();
    let value_moved_to_mask_bits = value << self.bits().trailing_zeros();

    byte_with_mask_bits_cleared | value_moved_to_mask_bits
  }

  pub fn with(&self, value: u8) -> u8 {
    value << self.bits().trailing_zeros()
  }

  /// Returns the value set at the mask bits.
  ///
  /// Applies [Mask] bits to byte and shifts everything right `n` times, where
  /// `n` is the number of trailing zeroes.
  ///
  /// Example; given:
  /// - A byte `0b1100_1100`
  /// - A mask `0b0000_1100`
  ///
  /// Then:
  /// - `mask.value(byte)` will produce `0b11`
  pub fn value(&self, byte: u8) -> u8 {
    let value_at_mask_bits = byte & self.bits();
    value_at_mask_bits >> self.bits().trailing_zeros()
  }

  pub fn is_set(&self, byte: u8) -> bool {
    self.value(byte) > 0
  }
}

impl Mask {
  pub fn exec_for(engine: Engine) -> Mask {
    match engine {
      Engine::E1 => Mask::ENGINE1_EXEC,
      Engine::E2 => Mask::ENGINE2_EXEC,
      Engine::E3 => Mask::ENGINE3_EXEC,
    }
  }

  pub fn mode_for(engine: Engine) -> Mask {
    match engine {
      Engine::E1 => Mask::ENGINE1_MODE,
      Engine::E2 => Mask::ENGINE2_MODE,
      Engine::E3 => Mask::ENGINE3_MODE,
    }
  }

  pub fn ratiometric_dimming_for(channel: Channel) -> Mask {
    match channel {
      Channel::D1 => Mask::D1_RATIO_EN,
      Channel::D2 => Mask::D2_RATIO_EN,
      Channel::D3 => Mask::D3_RATIO_EN,
      Channel::D4 => Mask::D4_RATIO_EN,
      Channel::D5 => Mask::D5_RATIO_EN,
      Channel::D6 => Mask::D6_RATIO_EN,
      Channel::D7 => Mask::D7_RATIO_EN,
      Channel::D8 => Mask::D8_RATIO_EN,
      Channel::D9 => Mask::D9_RATIO_EN,
    }
  }

  pub fn on_off_for(channel: Channel) -> Mask {
    match channel {
      Channel::D1 => Mask::D1_ON,
      Channel::D2 => Mask::D2_ON,
      Channel::D3 => Mask::D3_ON,
      Channel::D4 => Mask::D4_ON,
      Channel::D5 => Mask::D5_ON,
      Channel::D6 => Mask::D6_ON,
      Channel::D7 => Mask::D7_ON,
      Channel::D8 => Mask::D8_ON,
      Channel::D9 => Mask::D9_ON,
    }
  }
}


src/debug.rs
/// Macros for low level debugging when developing against I2C devices.

/// Create a debugging scope on the provided context.
///
/// A debugging scope causes any subsequent [`text!`] or [`byte!`] calls to be
/// printed with additional padding and automatically ends once the current code
/// scope ends.
///
/// Scopes can be nested, where the inner scope will further pad any output.
///
/// The provided context (`ctx`) can be any struct that has two fields
/// accessible to these macros:
/// 1. `debug_enabled: bool`
/// 2. `debug_depth: Rc<Cell<usize>>`
///
/// Example:
/// ```
/// debug::scope!(ctx, "entering scope 1");
/// debug::text!(ctx, "message 1")
/// debug::scope!(ctx, "entering scope 2");
/// debug::text!(ctx, "message 2")
/// {
///   debug::scope!(ctx, "entering scope 3");
///   debug::text!(ctx, "message 3")
/// } // scope 3 ends
/// debug::text!("message 4")
/// ```
/// Prints:
/// ```text
/// entering scope 1
///   message 1
///   entering scope 2
///     message 2
///     entering scope 3
///       message 3
///     message 4
/// ```
///
/// The nesting ability of scopes becomes really useful when composing I2C
/// operations; example:
/// ```
/// fn start() {
///   debug::scope!("start()");
///   let byte = 1;
///   debug::byte!(byte, "write start byte", byte)
///   device.write(byte);
/// }
///
/// fn stop() {
///   debug::scope!("stop()");
///   let byte = 0;
///   debug::byte!(byte, "write stop byte", byte)
///   device.write(byte);
/// }
///
/// fn restart() {
///   debug::scope!("restart()");
///   stop();
///   start();
/// }
///
/// restart();
/// // ...
/// stop();
/// ```
///
/// Would result in the output:
/// ```text
/// restart()
///   stop()
///     00000000 write stop byte
///   start()
///     00000001 write start byte
/// stop()
///   00000000 write stop byte
/// ```
#[macro_export]
macro_rules! scope {
  ($ctx:ident, $fmt:expr $(, $arg:expr)* $(,)?) => {
    // _unused is dropped at end of scope where this macro is called, which
    // decrements debug depth.
    let _unused = if $ctx.debug_enabled {
      let mut cur_depth = $ctx.debug_depth.lock().unwrap();
      let padding = "  ".repeat(*cur_depth);
      println!("{}{} {{", padding, format!($fmt $(, $arg)*));
      *cur_depth += 1;
      // lock releases after return

      // debug_depth ArcMutex needs to be cloned since guard() takes ownership
      // of the arg. Code in block below is executed at the end of the code
      // scope where the `scope` macro is called.
      Some(scopeguard::guard($ctx.debug_depth.clone(), |depth| {
        let mut cur_depth = depth.lock().unwrap();
        if *cur_depth > 0 {
          *cur_depth -= 1;
        }
        println!("{}}}", "  ".repeat(*cur_depth));
      }))
    } else {
      None
    };
  };
}
pub use scope;

/// Print a formated message at current debug depth.
///
/// Example:
/// ```
/// debug::text!(ctx, "1. top-level debug text");
/// debug::scope!(ctx, "2. in-scope");
/// debug::text!(ctx, "2a. in-scope debug text");
/// ```
///
/// Prints:
/// ```text
/// 1. top-level debug text
/// 2. in-scope
///   2a. in-scope debug text
/// ```
#[macro_export]
macro_rules! text {
  ($ctx:expr, $fmt:expr $(, $arg:expr)* $(,)?) => {
    if $ctx.debug_enabled {
      let padding = "  ".repeat(*$ctx.debug_depth.lock().unwrap());
      println!("{}{}", padding, format!($fmt $(, $arg)*));
    };
  };
}
pub use text;

/// Print the binary representation for the given byte, along with a formatted
/// description at the current debug depth.
///
/// Example:
/// ```
/// let value = 0b0010_1010;
/// debug::byte(value, "is binary for {}", value);
/// ```
/// Prints:
/// ```text
/// 00101010 is binary for 42
/// ```
#[macro_export]
macro_rules! byte {
  ($ctx:expr, $value:expr, $($description:tt)*) => {
    if $ctx.debug_enabled {
      let value: u8 = $value;
      let formatted_string = format!($($description)*);
      debug::text!(
        $ctx,
        "{:08b} {}",
        value,
        formatted_string,
      );
    }
  };
}
pub use byte;


src/types.rs
/// Output channels.
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum Channel {
  D1 = 0,
  D2,
  D3,
  D4,
  D5,
  D6,
  D7,
  D8,
  D9,
}

/// Master faders.
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum Fader {
  F1 = 0,
  F2,
  F3,
}

/// Programming engines.
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum Engine {
  E1 = 0,
  E2,
  E3,
}

/// Engine execution control modes.
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum EngineExec {
  Hold = 0,
  Step,
  Free,
  ExecuteOnce,
}

/// Engine modes (i.e. state).
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum EngineMode {
  Disabled = 0,
  LoadProgram,
  RunProgram,
  Halt,
}

/// Charge pump modes.
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum ChargePumpMode {
  Off = 0,
  Bypass,
  Boosted,
  Auto,
}

impl From<u8> for ChargePumpMode {
  fn from(value: u8) -> Self {
    match value {
      0b00 => Self::Off,
      0b01 => Self::Bypass,
      0b10 => Self::Boosted,
      0b11 => Self::Auto,
      _ => panic!("invalid value for ChargePumpMode {:b}", value),
    }
  }
}

/// IC clock selection.
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum ClockSelection {
  ForceExternal = 0,
  ForceInternal,
  Automatic,
  PreferInternal,
}

impl From<u8> for ClockSelection {
  fn from(value: u8) -> Self {
    match value {
      0b00 => Self::ForceExternal,
      0b01 => Self::ForceInternal,
      0b10 => Self::Automatic,
      0b11 => Self::PreferInternal,
      _ => panic!("invalid value for ClockSelection {:b}", value),
    }
  }
}

/// Miscellaneous settings.
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub struct Misc {
  /// EN_AUTO_INCR
  pub auto_increment_enabled: bool,
  /// POWERSAVE_EN
  pub powersave_enabled: bool,
  /// CHARGE_PUMP_EN
  pub charge_pump_mode: ChargePumpMode,
  /// PWM_PS_EN
  pub pwm_powersave_enabled: bool,
  /// CLK_DET_EN and INT_CLK_EN
  pub clock_selection: ClockSelection,
}


src/program.rs
use crate::Channel;

/// Maximum number of instructions supported by programming engine.
///
/// There are 6 pages, each with 16 instructions.
pub const MAX_INSTRUCTIONS: u8 = 96;
/// Number of instructions per page.
pub const INSTRUCTIONS_PER_PAGE: u8 = 16;
/// Number of pages supported by programming engine.
pub const MAX_PAGES: u8 = 6;
/// Number of variables supported by programming engine.
pub const MAX_VARS: u8 = 4;

/// Programming engine variables.
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
#[repr(u8)]
pub enum Variable {
  A = 0,
  B,
  C,
  D,
}

/// Ramp step time span.
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
#[repr(u8)]
pub enum PreScale {
  // Cycle time of 0.488ms
  CT0_488 = 0,
  // Cycle time of 15.625ms
  CT15_625 = 1,
}

/// Ramp direction.
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
#[repr(u8)]
pub enum Direction {
  Up = 0,
  Down = 1,
}

/// Representation for a programming engine instruction.
///
/// Refer to spec sections 7.6.3 through 7.6.7
pub struct Instruction {
  pub msb: u8,
  pub lsb: u8,
}

impl Instruction {
  /// Word (u16) representation for an instruction.
  pub fn as_u16(&self) -> u16 {
    (self.msb as u16) << 8 | self.lsb as u16
  }
}

impl From<u16> for Instruction {
  /// Convert a word (u16) to an [Instruction].
  fn from(value: u16) -> Self {
    Self {
      msb: ((value >> 8) & 0xFF) as u8,
      lsb: (value & 0xFF) as u8,
    }
  }
}

impl Instruction {
  // Driver instructions

  pub fn ramp(
    cycle_time: PreScale,
    cycles_per_step: u8,
    direction: Direction,
    number_of_steps: u8,
  ) -> Self {
    let mut msb = cycles_per_step << 1; // TODO check bounds
    msb |= (cycle_time as u8) << 6;
    msb |= direction as u8;

    Self {
      msb,
      lsb: number_of_steps,
    }
  }

  pub fn ramp_from_vars(
    pre_scale: bool,
    ascending: bool,
    step_time_var: Variable,
    increments_var: Variable,
  ) -> Self {
    let mut lsb = ((step_time_var as u8) << 2) | (increments_var as u8);
    if pre_scale {
      lsb |= 1 << 6;
    }
    if ascending {
      lsb |= 1 << 5;
    }
    Self {
      msb: 0b1000_0100,
      lsb,
    }
  }

  pub fn set_pwm(value: u8) -> Self {
    Self {
      msb: 0b0100_0000,
      lsb: value,
    }
  }

  pub fn set_pwm_from_var(var: Variable) -> Self {
    Self {
      msb: 0b1000_0100,
      lsb: 0b0110_0000 | (var as u8),
    }
  }

  pub fn wait(cycle_time: PreScale, cycles: u8) -> Self {
    let mut msb = cycles << 1; // TODO check bounds
    msb |= (cycle_time as u8) << 6;
    Self {
      msb,
      lsb: 0b0000_0000,
    }
  }

  // Mapping instructions

  /// Create LED engine-to-LED mapping instruction.
  ///
  /// Associates the supplied [channels](Channel) with the the active engine.
  /// This information is not present in the main spec, but can be found in the
  /// [LP55231 evaluation kit](https://www.ti.com/lit/ug/snvu214b/snvu214b.pdf)
  /// User's Guide, on page 23.
  ///
  /// |Bit    |15|14|13|12|11|10|09|08|07|06|05|04|03|02|01|00|
  /// |-------|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|
  /// |Channel| -| -| -| -| -| -| -|D9|D8|D7|D6|D5|D4|D3|D2|D1|
  pub fn map_channels(channels: &[Channel]) -> Self {
    let mut map_bits = 0b0000_0000_0000_0000;
    for channel in channels.iter() {
      map_bits |= 1 << (*channel as u8);
    }
    Self::from(map_bits)
  }

  pub fn mux_ld_start(sram_address: u8) -> Self {
    Self {
      msb: 0b1001_1110,
      lsb: check_addr(sram_address),
    }
  }

  pub fn mux_map_start(sram_address: u8) -> Self {
    Self {
      msb: 0b1001_1100,
      lsb: check_addr(sram_address),
    }
  }

  pub fn mux_ld_end(sram_address: u8) -> Self {
    Self {
      msb: 0b1001_1100,
      lsb: 0b1000_0000 | check_addr(sram_address),
    }
  }

  pub fn mux_sel(led_select: u8) -> Self {
    Self {
      msb: 0b1001_1101,
      lsb: led_select,
    }
  }

  pub fn mux_clr() -> Self {
    Self {
      msb: 0b1001_1101,
      lsb: 0b0000_0000,
    }
  }

  pub fn mux_map_next() -> Self {
    Self {
      msb: 0b1001_1101,
      lsb: 0b1000_0000,
    }
  }

  pub fn mux_map_prev() -> Self {
    Self {
      msb: 0b1001_1101,
      lsb: 0b1100_0000,
    }
  }

  pub fn mux_ld_next() -> Self {
    Self {
      msb: 0b1001_1101,
      lsb: 0b1000_0001,
    }
  }

  pub fn mux_ld_prev() -> Self {
    Self {
      msb: 0b1001_1101,
      lsb: 0b1100_0001,
    }
  }

  pub fn mux_ld_addr(sram_address: u8) -> Self {
    Self {
      msb: 0b1001_1111,
      lsb: check_addr(sram_address),
    }
  }

  pub fn mux_map_addr(sram_address: u8) -> Self {
    Self {
      msb: 0b1001_1111,
      lsb: 0b1000_0000 | check_addr(sram_address),
    }
  }

  // Branch instructions

  pub fn rst() -> Self {
    Self {
      msb: 0b0000_0000,
      lsb: 0b0000_0000,
    }
  }

  pub fn branch(step_number: u8, loop_count: u8) -> Self {
    let mut bits: u16 = 0b1010_0000_0000_0000;
    bits |= step_number as u16; // TODO validate bounds
    bits |= (loop_count as u16) << 7; // TODO validate bounds
    Self::from(bits)
  }

  pub fn branch_vars(step_number: u8, loop_count_var: Variable) -> Self {
    let mut bits: u16 = 0b1000_0110_0000_0000;
    bits |= loop_count_var as u16;
    bits |= (step_number as u16) << 2; // TODO check bounds
    Self::from(bits)
  }

  pub fn int() -> Self {
    Self {
      msb: 0b1100_0100,
      lsb: 0b0000_0000,
    }
  }

  pub fn end(interrupt: bool, reset_program_counter: bool) -> Self {
    let mut msb = 0b1100_0000;
    if interrupt {
      msb |= 1 << 4;
    }
    if reset_program_counter {
      msb |= 1 << 3
    }
    Self {
      msb,
      lsb: 0b0000_0000,
    }
  }

  pub fn jne(
    num_instructions_to_skip: u8,
    var_1: Variable,
    var_2: Variable,
  ) -> Self {
    let mut instr: u16 = 0b1000_1000_0000_0000;
    instr |= var_1 as u16;
    instr |= (var_2 as u16) << 2;
    instr |= (num_instructions_to_skip as u16) << 4; // TODO check bounds.

    Self::from(instr)
  }

  pub fn jl(
    num_instructions_to_skip: u8,
    var_1: Variable,
    var_2: Variable,
  ) -> Self {
    let mut instr: u16 = 0b1000_1010_0000_0000;
    instr |= var_2 as u16;
    instr |= (var_1 as u16) << 2;
    instr |= (num_instructions_to_skip as u16) << 4; // TODO check bounds.

    Self::from(instr)
  }

  pub fn jge(
    num_instructions_to_skip: u8,
    var_1: Variable,
    var_2: Variable,
  ) -> Self {
    let mut instr: u16 = 0b1000_1100_0000_0000;
    instr |= var_2 as u16;
    instr |= (var_1 as u16) << 2;
    instr |= (num_instructions_to_skip as u16) << 4; // TODO check bounds.

    Self::from(instr)
  }

  pub fn je(
    num_instructions_to_skip: u8,
    var_1: Variable,
    var_2: Variable,
  ) -> Self {
    let mut instr: u16 = 0b1000_1110_0000_0000;
    instr |= var_2 as u16;
    instr |= (var_1 as u16) << 2;
    instr |= (num_instructions_to_skip as u16) << 4; // TODO check bounds.

    Self::from(instr)
  }

  pub fn ld(target_var: Variable, value: u8) -> Self {
    Self {
      msb: 0b1001_0000 | ((target_var as u8) << 2),
      lsb: value,
    }
  }

  pub fn add_numerical(target_var: Variable, value: u8) -> Self {
    Self {
      msb: 0b1001_0001 | ((target_var as u8) << 2),
      lsb: value,
    }
  }

  pub fn add_vars(
    target_var: Variable,
    var_1: Variable,
    var_2: Variable,
  ) -> Self {
    Self {
      msb: 0b1001_0011 | ((target_var as u8) << 2),
      lsb: ((var_1 as u8) << 2) | (var_2 as u8),
    }
  }

  pub fn sub_numerical(target_var: Variable, value: Variable) -> Self {
    Self {
      msb: 0b1001_0010 | ((target_var as u8) << 2),
      lsb: value as u8,
    }
  }

  pub fn sub_vars(
    target_var: Variable,
    var_1: Variable,
    var_2: Variable,
  ) -> Self {
    Self {
      msb: 0b1001_0011 | ((target_var as u8) << 2),
      lsb: 0b0001_0000 | ((var_1 as u8) << 2) | (var_2 as u8),
    }
  }
}

fn check_addr(addr: u8) -> u8 {
  if addr > MAX_INSTRUCTIONS {
    panic!(
      "invalid sram_address {} - max is {}",
      addr, MAX_INSTRUCTIONS
    )
  }
  addr
}
